meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+				end: `#${controlForPiece.id}.end - ${piece.postRollDuration ?? 0}`,
+			},
+			layer: '',
+		}
+	}
+
+	private shouldPieceHavePreRollGroup(controlForPiece: TimelineObjectControl, piece: Piece): boolean {
+		// TODO: If we can't get rid of start="now" then we need to check for "now" here.
+		return (controlForPiece.enable as TimelineEnable).start === 0 && piece.preRollDuration > 0
+	}
+
+	private createPreRollGroupForControl(
+		controlObject: TimelineObjectControl,
+		parentGroup: TimelineObjectGroup
+	): TimelineObjectControl {
+		return {
+			id: `${PIECE_PRE_ROLL_PREFIX}${controlObject.id}`,
+			enable: {
+				start: `#${parentGroup.id}.start`,
+			},
+			layer: '',
+			content: {
+				type: TimelineObjectType.CONTROL,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+	}
+
+	private mapToTimelineObjectForPieceGroup(
+		timelineObject: TimelineObject,
+		childGroupForPiece: TimelineObjectGroup,
+		piece: Piece
+	): TimelineObject {
+		// TODO: Core checks "HoldMode" on the timelineObject - ignore for now
+		const timelineObjectCopy: TimelineObject = JSON.parse(JSON.stringify(timelineObject))
+		timelineObjectCopy.id = `${childGroupForPiece.id}_${piece.id}_${timelineObject.id}`
+		timelineObjectCopy.inGroup = childGroupForPiece.id
+		return timelineObjectCopy
+	}
+
+	private createTimelineObjectGroupForPreviousPart(
+		previousPart: Part,
+		activePart: Part,
+		activeGroup: TimelineObjectGroup
+	): TimelineObjectGroup {
+		// TODO: This if-statement shouldn't be necessary in production
+		if (previousPart.getExecutedAt() <= 0) {
+			throw new Error(
+				`Previous Part: ${previousPart.name} does not have a valid "executedAt" - something went wrong when setting the previous Part.`
+			)
+		}
+
+		const previousGroup: TimelineObjectGroup = {
+			id: `${PREVIOUS_GROUP_PREFIX}${previousPart.id}`,
+			priority: LOW_PRIORITY,
+			isGroup: true,
+			children: [],
+			enable: {
+				start: previousPart.getExecutedAt(),
+				// end: `#${activeGroup.id}.start`
+				end: `#${activeGroup.id}.start + ${activePart.getTimings().previousPartContinueIntoPartDuration}`,
+			},
+			layer: '',
+			content: {
+				type: TimelineObjectType.GROUP,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+
+		previousGroup.children = previousPart
+			.getPiecesWithLifespan([PieceLifespan.WITHIN_PART])
+			.flatMap((piece) => this.generateGroupsAndTimelineObjectsForPiece(piece, previousPart, previousGroup))
+
+		return previousGroup
+	}
+
+	private createTimelineObjectGroupsForInfinitePieces(rundown: Rundown): TimelineObjectGroup[] {
+		const activePart: Part = rundown.getActivePart()
+		const infinitePieceTimelineObjectGroups: TimelineObjectGroup[] = []
+		rundown
+			.getInfinitePieces()
+			.filter(
+				(piece) =>
+					piece.transitionType ===
+					TransitionType.NO_TRANSITION /* TODO: && filter for not disabled - if that becomes a thing */
+			)
+			.filter((piece) => piece.partId !== activePart.id)
+			.forEach((piece) => {
```

Comment:
The function passed to this `forEach` should propably be extracted to its own method.

Reply:
It could, but since it does three things it _might_ make more sense to extract those things and then call them here (and might also make sense to change the foreach to a map)
```
.map((piece) => {
  this.assertExecutedAt(piece)
  const infiniteGroup = this.createInfinitePieceGroup(piece)
  infiniteGroup.children = this.createTimelineObjectsForPieceGroup(infiniteGroup, piece)
  return infiniteGroup
})
```